classes/privacy/provider.php

```diff
@@ -0,0 +1,266 @@
+                 WHERE kc_users.userid = :userid
+        ";
+        $params = [
+            'userid' => $userid,
+        ];
+        $details = $DB->get_records_sql($sql, $params);
+
+        // Export the course related user information.
+        foreach ($contextlist->get_contexts() as $context) {
+            $courseid = $context->instanceid; // Tentative, as also need to check contextlevel.
+            if ($context->contextlevel === CONTEXT_COURSE && array_key_exists($courseid, $details)) {
+                $data = new \stdClass();
+                $data->moodleuserid = $details[$courseid]->userid;
+                $data->kuracloudstudentid = $details[$courseid]->remote_studentid;
+                // studentid is within instance and course, but not much value in adding them to exported data.
+                writer::with_context($context)->export_data(['kuraCloud'], $data);
+            }
+        }
+    }
+
+
+    /**
+     * Delete all personal data for all users in the specified context.
+     *
+     * @param context $context Context to delete data from.
+     */
+    public static function delete_data_for_all_users_in_context(\context $context) {
+        if ($context->contextlevel !== CONTEXT_COURSE) {
+            return;
+        }
+
+        global $DB;
+        $mapping = $DB->get_record('block_kuracloud_courses', array('courseid' => $context->instanceid));
+        if ($mapping) {
+            $DB->delete_records('block_kuracloud_users', array('remote_instanceid' => $mapping->remote_instanceid, 'remote_courseid' => $mapping->remote_courseid));
+        }
+    }
+
+
+    /**
+     * Delete all user data for the specified user, in the specified contexts.
+     *
+     * @param  approved_contextlist $contextlist The approved contexts to export information for.
+     */
+    public static function delete_data_for_user(approved_contextlist $contextlist) {
+        $userid = $contextlist->get_user()->id;
+        kuracloud_delete_stale_data_for_user($userid);
+
+        if (empty($contextlist->count())) {
+            return;
+        }
+
+        global $DB;
+
+        // Get a map from courseid to the related user record.
+        $sql = "SELECT kc_courses.courseid, kc_users.id
+                  FROM {block_kuracloud_courses} kc_courses
+            INNER JOIN {block_kuracloud_users} kc_users on kc_users.remote_instanceid = kc_courses.remote_instanceid AND kc_users.remote_courseid = kc_courses.remote_courseid
+                 WHERE kc_users.userid = :userid
+        ";
+        $params = [
+            'userid' => $userid,
+        ];
+        $details = $DB->get_records_sql($sql, $params);
+
+        foreach ($contextlist->get_contexts() as $context) {
+            $courseid = $context->instanceid; // Tentative, as also need to check contextlevel.
+            if ($context->contextlevel === CONTEXT_COURSE && array_key_exists($courseid, $details)) {
```

Comment:
Can't straight up delete this as this function call is needed? `array_key_exists`

Reply:
Yes, and was easier and more in style of other routines to loop than try and build into sql query.